Blog: When to use waitUntil (#781)
inngest/website: pages/blog/_posts/vercel-cloudflare-wait-until.mdx

---

> heading: When to use Vercel and Cloudflare's waitUntil utility

Reviewer: I feel like this title might be a bit authoritative. Maybe: "What is waitUntil (Vercel, Cloudflare) and when do I need it?"

subtitle: What is it, when to use it, and when not to use it

> image: /assets/blog/vercel-cloudflare-wait-until/featured-image.png

Reviewer: I like the code as a picture. What do you think about making the font bigger and maybe adding some short title or question? Just so it will be contextualized better.

(I can do it)

date: 2024-05-16
author: Dan Farrelly
---

> 
> Recently, [Vercel announced](https://vercel.com/changelog/waituntil-is-now-available-for-vercel-functions) that their `waitUntil` utility method is now available for all serverless functions. Now, this useful feature is available for both Vercel's Node.js and edge functions. A similar utility is also available for [Cloudflare Workers](https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil).

Reviewer: ```suggestion

Recently, [Vercel announced](https://vercel.com/changelog/waituntil-is-now-available-for-vercel-functions) that their `waitUntil` utility method is now available for all serverless functions. Now, this useful feature is available for both Vercel's Node.js and edge functions. This utility is not unique to Vercel -- [Cloudflare Workers](https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil) also offers a similar one.
```

> This is a very useful tool, but it's important to understand when to use it and when not to use it. In this post, I'll explain:

Reviewer: ```suggestion
This is a very useful tool so let's dive into what `waitUntil` is and what it does, when to use it, and when it's better not to.
```

Changing it so it's sounds more like you are excited about it, and not telling others to be that.

## What does `waitUntil` do

> To understand what `waitUntil` does, we need to understand how serverless functions work. This also helps us understand why it exists.

Reviewer: ```suggestion
To understand what `waitUntil` does and why it exists, let's talk about how serverless functions work.
```

With JavaScript serverless functions, you have to `await` all asynchronous operations to ensure that the function doesn't return before the asynchronous operation is complete. If you don't `await` an asynchronous operation, the serverless function might be shut down before the operation is complete. If this happens, the operation might be canceled or fail which results in unpredictable behavior. This is why it's important to `await` all asynchronous operations in serverless functions.

```js
// ❌ BAD: If sendMetrics is an async function (a Promise), there is no guarantee it will succeed
async function POST(request) {
  const data = await createPost(request.body)
  sendMetrics({ metric: 'posts.created', increment: 1 })
  return new Response(data)
}
// ✅ GOOD: Awaiting will ensure it completes before the function returns
async function POST(request) {
  const data = await createPost(request.body)
  await sendMetrics({ metric: 'posts.created', increment: 1 })
  return new Response(data)
}
```

> The downside of this is that it can slow down the response time of your serverless functions. You ideally want to [keep the critical path of your request minimal](https://www.epicweb.dev/talks/improve-performance-and-reliability-of-your-api-with-events-and-background-functions) so you return the fastest response possible for your users. This is where `waitUntil` comes in.

Reviewer: ```suggestion
The downside of using `await` is that it can slow down the response time of your serverless functions. You ideally want to [keep the critical path of your request minimal](https://www.epicweb.dev/talks/improve-performance-and-reliability-of-your-api-with-events-and-background-functions) so you return the fastest response possible for your users. This is where `waitUntil` comes in.
```

`waitUntil` allows you to ensure that an async function (a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)) completes before the serverless function is shut down. As the example code above doesn't require any data returned from `sendMetrics`, we can use `waitUntil` to ensure that the async function completes in the background without blocking the response from returning to the user.

```js
// ✅ GOOD: The sendMetrics function will execute in the background,
// but not block the response from returning to the user
async function POST(request) {
  const data = await createPost(request.body)
  waitUntil(sendMetrics({ metric: 'posts.created', increment: 1 }))
  return new Response(data)
}
```

> `waitUntil` is a very useful tool that ensures completion, but it does not ensure _success_. If the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. It's important to understand when to use it and when to not use it.

Reviewer: ```suggestion
So, is `waitUntil` a solution to all possible problems? Well, it is a very useful tool and it ensures completion. However, it does not ensure _success_. For example, if the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. So when should we use it?
```

## When to use `waitUntil`

> `waitUntil` is useful when you need to perform some asynchronous work that is not critical to the response of the serverless function. This is useful for things like sending metrics, logging, or other things that should not block the user's response. The user should not pay the penalty for these operations, so they can be done in the background.

Reviewer: ```suggestion
`waitUntil` is useful when you need to perform some asynchronous work that is not critical to the response of the serverless function. It is great for things like sending metrics, logging, or other things that should not block the user's response. The user should not pay the penalty for these operations, so they can be done in the background.
```

## When _not_ to use `waitUntil`

> As mentioned above, `waitUntil` does not ensure _success_. If the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. This is important to understand when deciding when to use `waitUntil`.

Reviewer: ```suggestion
As I mentioned above, `waitUntil` does not ensure _success_. If the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. Let's keep that in mind when deciding is to use `waitUntil`.
```

> You shouldn't use `waitUntil` if the asynchronous operation effects critical business logic in your application. This is what makes it great for logging and metrics. If the operation fails, it's not the end of the world. But if the operation is critical to the response of the serverless function, you should `await` it instead.

Reviewer: ```suggestion
In other words, `waitUntil` is not the tool for the job if the asynchronous operation affects critical business logic in your application (which is precisely what makes it great for logging and metrics!). If the operation fails, it's not the end of the world. But if the operation is critical to the response of the serverless function, `await` is the way to go.
```

Making it sound less authoritative.

## Can I run background jobs with `waitUntil`?

Short answer: No.

Long answer: `waitUntil` is not designed for running background jobs. There are no retries or tools for handling failures. For running background jobs, you should use a system that handles queueing, retries, and logs failures.

> Inngest is designed to handle background jobs reliably. It automatically handles retries, logs, and failures. If you need to run background jobs, you should use Inngest instead of `waitUntil`.

Reviewer: ```suggestion
Inngest is designed to handle background jobs reliably. It automatically handles retries, logs, and failures. If you need to run background jobs, you should consider using Inngest instead of `waitUntil`.
```

```js
import { inngest } from "src/inngest";

async function POST(request) {
  const data = await createPost(request.body)
  waitUntil(inngest.send({ name: "post.created", data: data }))
  return new Response(data)
}
```

Typically, Inngest functions contain core business logic, so you need to ensure that the functions are triggered successfully. The key question is the reliability of `waitUntil` and the function it calls. Let's break down some things to consider:

* The Inngest Event API has very high uptime and is designed to be reliable.
* The `inngest.send()` function is very simple - mostly just an `POST` request to the Event API.
* The `waitUntil` utility from Vercel and Cloudflare is designed to be reliable.
* Networking issues or blips can always cause issues with outbound requests from any platform.

Given these points, it is safe to use `waitUntil` with `inngest.send()` with the caveat that if there is a networking issue, the event may not reach Inngest. While this is a rare occurrence, it is a risk to be aware of and is a trade off that you need to decide for yourself.

> If you don't like this risk, you can use `inngest.send()` without `waitUntil`, but it will be blocking:

Reviewer: ```suggestion
If you don't like to take risk, you can use `inngest.send()` without `waitUntil`, but it will be blocking:
```
